### PR TITLE
Periodic padding

### PR DESCRIPTION
**Pull request by jder**
NB Stacked on #407 (and so #406).

We have had a clear artifact in our outputs at lon=0. I believe there were 2 causes ([thread](https://openathena.slack.com/archives/C09FXFDR2BF/p1758734811767269)):
* On 1 degree data: since there are 4 ConvNext blocks and each has a 2x downsample before them, that gives a final image width of 360/(2^4) which is 22.5. So, the last downsample block (from 45 -> 22) will end up sampling from implicit zero padding in the far-right column (which is lon=359). And on the upscale from 22 -> 45 we pad with zeros. These are both non-periodic.
* In any dataset, the bilinear upsample was non-periodic i.e. `[[1, 2, 3], ...]` gets upsampled to `[[1.0, 1.25, 1.75, 2.25, 2.75, 3.0], ...]` rather than circularly to eg `[[1.0, 1.5, 2.0, 2.5, 3.0, 2.0], …]`. 

This PR resolves the latter and appears to resolve the artifact for us on half-degree data. 
Before (from [om4_samudra_halfdeg_wide_dmse_no_cap](https://wandb.ai/m2lines/ocean-emulators/runs/k5szfus3?nw=nwuseroa_jder)):
SST_map_snapshot_t_0_before

After (from [periodic_test](https://wandb.ai/m2lines/ocean-emulators/runs/6dtu2st0?nw=nwuseroa_jder) and [periodic_eval](https://wandb.ai/m2lines/ocean-emulators/runs/cnmoqz0c?nw=nwuseroa_jder)):
SST_map_snapshot_t_0_after

### CODE REVIEW

File: src/ocean_emulators/config.py [hunk 1]
```diff
@@ -322,7 +323,7 @@ def build(self, in_channels: int, out_channels: int) -> PerceiverEncoder:
 
 
 DownSamplingBlocks = Literal["avg_pool", "max_pool"]
-UpSamplingBlocks = Literal["bilinear_upsample", "transposed_conv"]
+UpSamplingBlocks = Literal["bilinear_upsample", "transposed_conv", "periodic_upsample"]
```
Comment (adam-subel) [inline src/ocean_emulators/config.py]: From a naming perspective, would labeling this "zonally_periodic_upsample" be more specific?

File: src/ocean_emulators/models/modules/blocks.py [hunk 2]
```diff
@@ -46,6 +46,45 @@ def forward(self, x):
         return self.upsampler(x)
 
 
+class PeriodicBilinearUpsample(torch.nn.Module):
```
Comment (adam-subel) [inline src/ocean_emulators/models/modules/blocks.py]: Same comment as before, may be worth changing the name to clarify this is only zonally periodic

File: tests/test_periodic.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+@pytest.mark.parametrize(
+    "upsampler_cls",
+    [
+        pytest.param(BilinearUpsample, id="bilinear"),
+        pytest.param(PeriodicBilinearUpsample, id="periodic"),
+    ],
+)
+@pytest.mark.parametrize("width", [8, 10, 16])
+@pytest.mark.parametrize("num_blocks", [1, 2, 3])
+def test_downscale_then_upscale_rotation_invariance(
```
Comment (adam-subel) [inline tests/test_periodic.py]: Another nomenclature question. If I am not mistaken, this is a translation test not a rotation test, maybe the naming should adjust to reflect that

**Review by adam-subel — APPROVED** [review]
Looks good, I left a couple suggestions for naming, but I will let you decide if any of it is worth changing.

**Comment by jder** [comment]
Makes sense to me, thanks for the review!

File: src/ocean_emulators/config.py [hunk 4]
```diff
@@ -354,6 +357,8 @@ def create_upsampling_block(in_channels: int, out_channels: int):
                     return TransposedConvUpsample(
                         in_channels=in_channels, out_channels=out_channels
                     )
+                case "zonally_periodic_upsample":
+                    return ZonallyPeriodicBilinearUpsample()
```
Comment (alxmrs) [inline src/ocean_emulators/config.py:361]: ~I'm curious why these don't need `in_channels` and `out_channels`.~

It's cool that this sampling is independent of the inputs and outputs.

File: src/ocean_emulators/models/modules/blocks.py [hunk 5]
```diff
@@ -46,6 +46,49 @@ def forward(self, x):
...
+                raise ValueError(
+                    "ZonallyPeriodicBilinearUpsample only supports 2x upsampling"
+                )
+            self.scale_h, self.scale_w = upsampling
+        else:
+            if upsampling != 2:
+                raise ValueError(
+                    "ZonallyPeriodicBilinearUpsample only supports 2x upsampling"
+                )
+            self.scale_h = self.scale_w = upsampling
```
Comment (alxmrs) [inline src/ocean_emulators/models/modules/blocks.py]: 🐑  optional alternative
```suggestion
        if isinstance(upsampling, int):
            upsampling = (upsampling, upsampling)
        if tuple(upsampling) != (2, 2):
            raise ValueError(
                    "ZonallyPeriodicBilinearUpsample only supports 2x upsampling"
            )
        self.scale_h, self.scale_w = upsampling
```

File: src/ocean_emulators/models/modules/blocks.py [hunk 6]
```diff
@@ -46,6 +46,49 @@ def forward(self, x):
...
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # First upsample along latitude/height using standard "bilinear" interpolation.
+        # (This is just linear interpolation but pytorch requires us to use "bilinear" for
+        # 2d data.)
```
Comment (alxmrs) [inline src/ocean_emulators/models/modules/blocks.py:65]: Thanks, these notes are helpful.

File: src/ocean_emulators/models/modules/blocks.py [hunk 7]
```diff
@@ -46,6 +46,49 @@ def forward(self, x):
...
+        right = torch.roll(up_height, shifts=-1, dims=-1)
+        midpoint = 0.5 * (left + right)
+
+        out = torch.empty(
+            (*up_height.shape[:-1], up_height.shape[-1] * self.scale_w),
+            dtype=up_height.dtype,
+            device=up_height.device,
+        )
+        out[..., ::2] = left
+        out[..., 1::2] = midpoint
```
Comment (alxmrs) [inline src/ocean_emulators/models/modules/blocks.py:83]: A really elegant upsample! :)

File: src/ocean_emulators/models/modules/blocks.py [hunk 8]
```diff
@@ -46,6 +46,49 @@ def forward(self, x):
...
+        left = up_height
+        right = torch.roll(up_height, shifts=-1, dims=-1)
+        midpoint = 0.5 * (left + right)
+
+        out = torch.empty(
+            (*up_height.shape[:-1], up_height.shape[-1] * self.scale_w),
+            dtype=up_height.dtype,
+            device=up_height.device,
+        )
+        out[..., ::2] = left
```
Comment (alxmrs) [inline src/ocean_emulators/models/modules/blocks.py:82]: Is the last dim always the channel dim? I thought it was dim=2 in data space? How does this apply to our CNNs, I think I'm a bit confused/curious.

File: tests/test_periodic.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+import torch
+
+from ocean_emulators.models.modules.blocks import (
+    AvgPool,
+    BilinearUpsample,
+    ZonallyPeriodicBilinearUpsample,
+)
+
+
+def _pad_like_unet(feature: torch.Tensor, skip: torch.Tensor) -> torch.Tensor:
```
Comment (alxmrs) [inline tests/test_periodic.py:12]: 🐑 What do you think of extracting the main unet pad block (above) into a function? I could go either way.

File: tests/test_periodic.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+@pytest.mark.parametrize(
+    "upsampler_cls",
+    [
+        pytest.param(BilinearUpsample, id="bilinear"),
+        pytest.param(ZonallyPeriodicBilinearUpsample, id="zonally_periodic"),
+    ],
+)
+@pytest.mark.parametrize("width", [8, 10, 16])
+@pytest.mark.parametrize("num_blocks", [1, 2, 3])
+def test_downscale_then_upscale_translation_invariance(
```
Comment (alxmrs) [inline tests/test_periodic.py:36]: 🐑 Seems fuzzable with hypothesis!

File: tests/test_periodic.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        # if the final plane would be a width of 1, there's no shift we can try
+        # below to confirm translation invariance
+        pytest.skip("Width is too small for the given number of blocks")
+    divisibility_condition = width % total_downscale == 0
+    batch, channels, height = 1, 3, 8
+    base_input = torch.arange(
+        batch * channels * height * width, dtype=torch.float32
+    ).reshape(batch, channels, height, width)
+
+    def run_pipeline(original: torch.Tensor, count) -> torch.Tensor:
```
Comment (alxmrs) [inline tests/test_periodic.py:56]: What a fun use of recursion!

File: tests/test_periodic.py (hunk 10 above)
Comment (alxmrs) [reply tests/test_periodic.py:36]: This is a well written test. Thanks for illustrating the difference.

**Review by alxmrs — APPROVED** [review]
Small nits, in general, thanks for the fix!

File: src/ocean_emulators/models/modules/blocks.py (hunk 8 above)
Comment (jder) [reply src/ocean_emulators/models/modules/blocks.py:82]: The last dim is longitude, which is what we want to up-sample here.
